README.md
# tcp-base

A base class for tcp client with basic functions

[![NPM version][npm-image]][npm-url]
[![Node.js CI](https://github.com/node-modules/tcp-base/actions/workflows/nodejs.yml/badge.svg)](https://github.com/node-modules/tcp-base/actions/workflows/nodejs.yml)
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/tcp-base.svg?style=flat-square
[npm-url]: https://npmjs.org/package/tcp-base
[codecov-image]: https://codecov.io/gh/node-modules/tcp-base/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/node-modules/tcp-base
[download-image]: https://img.shields.io/npm/dm/tcp-base.svg?style=flat-square
[download-url]: https://npmjs.org/package/tcp-base

## Install

```bash
npm install tcp-base --save
```

Node.js >= 6.0.0 required

## Usage

A quick guide to implement a tcp echo client

Client:

```js
'use strict';

const TCPBase = require('tcp-base');

/**
 * A Simple Protocol:
 *   (4B): request id
 *   (4B): body length
 *   ------------------------------
 *   body data
 */
class Client extends TCPBase {
  getHeader() {
    return this.read(8);
  }

  getBodyLength(header) {
    return header.readInt32BE(4);
  }

  decode(body, header) {
    return {
      id: header.readInt32BE(0),
      data: body,
    };
  }

  // heartbeat packet
  get heartBeatPacket() {
    return new Buffer([ 255, 255, 255, 255, 0, 0, 0, 0 ]);
  }
}

const client = new Client({
  host: '127.0.0.1',
  port: 8080,
});

const body = new Buffer('hello');
const data = new Buffer(8 + body.length);
data.writeInt32BE(1, 0);
data.writeInt32BE(body.length, 4);
body.copy(data, 8, 0);

client.send({
  id: 1,
  data,
  timeout: 5000,
}, (err, res) => {
  if (err) {
    console.error(err);
  }
  console.log(res.toString()); // should echo 'hello'
});
```

Server:

```js
'use strict';

const net = require('net');

const server = net.createServer(socket => {
  let header;
  let bodyLen;

  function readPacket() {
    if (bodyLen == null) {
      header = socket.read(8);
      if (!header) {
        return false;
      }
      bodyLen = header.readInt32BE(4);
    }

    if (bodyLen === 0) {
      socket.write(header);
    } else {
      const body = socket.read(bodyLen);
      if (!body) {
        return false;
      }
      socket.write(Buffer.concat([ header, body ]));
    }
    bodyLen = null;
    return true;
  }

  socket.on('readable', () => {
    try {
      let remaining = false;
      do {
        remaining = readPacket();
      }
      while (remaining);
    } catch (err) {
      console.error(err);
    }
  });
});
server.listen(8080);
```

[MIT](LICENSE)


History.md

3.1.1 / 2022-09-01
==================

**fixes**
  * [[`be4eee2`](http://github.com/node-modules/tcp-base/commit/be4eee2c8bc3f96b6de11639c5e346a08e92fdf0)] - fix: socket.on('error') instead of once (#13) (hyj1991 <<yeekwanvong@gmail.com>>)

**others**
  * [[`d0a814a`](http://github.com/node-modules/tcp-base/commit/d0a814a17da2d5dc910e3eee3706646972d9396c)] - 🤖 TEST: Run ci on GitHub Action (#14) (fengmk2 <<fengmk2@gmail.com>>)

3.1.0 / 2017-12-01
==================

**features**
  * [[`e59388a`](http://github.com/node-modules/tcp-base/commit/e59388a564304803d0b85222bce1dc3e945f6fac)] - feat: support unix-domain-socket (#11) (fisher <<fishbar@users.noreply.github.com>>)

**others**
  * [[`f9c56ee`](http://github.com/node-modules/tcp-base/commit/f9c56ee64afb38b4062e5f1a247db45cbce4192c)] - chore: bump to 3.0.0 (xiaochen.gaoxc <<xiaochen.gaoxc@alibaba-inc.com>>),

3.0.0 / 2017-04-20
==================

  * fix: invoke oneway after socket closed (#7)
  * refactor: [BREAKING-CHANGE] not support reconnect

2.0.0 / 2017-02-17
==================

  * refactor: [BREAKING-CHANGE] upgrade sdk-base (#5)

1.1.0 / 2016-11-30
==================

  * feat: add heartbeat timeout server host
  * feat: add socket write status

1.0.2 / 2016-11-11
==================

  * fix: use heartbeat logic to handle that connection stuck in CLOSE_WAIT status (#3)

1.0.1 / 2016-10-20
==================

  * fix: callback exception should not crash tcp-base (#2)

1.0.0 / 2016-09-18
==================

  * feat: implement tcp-base


test/error.test.js
'use strict';

const TCPBase = require('../');
const server = require('./support/server_immidiate_end');

describe('test/error.test.js', () => {
  class Client extends TCPBase {
    constructor(options) {
      Object.assign(options, {
        headerLength: 8,
        heartbeatInterval: 3000,
      });

      super(options);
    }

    getBodyLength(header) {
      return header.readInt32BE(0);
    }

    decode(buf, header) {
      let data;
      if (buf) {
        data = JSON.parse(buf);
      }
      return {
        id: header.readInt32BE(4),
        data,
      };
    }

    get heartBeatPacket() {
      return makeRequest(1).data;
    }
  }

  function makeRequest(id, content, oneway) {
    const header = Buffer.alloc(8);
    header.fill(0);
    const body = Buffer.from(JSON.stringify(content || {
      id,
      message: 'hello',
    }));
    header.writeInt32BE(body.length, 0);
    header.writeInt32BE(id, 4);
    return {
      id,
      oneway,
      data: Buffer.concat([ header, body ]),
    };
  }

  describe('should not throw uncaughtExeception', () => {
    it('when socket has been destroyed', done => {
      server.start(9090, () => {
        const client = new Client({
          host: '127.0.0.1',
          port: 9090,
        });

        client.once('connect', () => {
          client.send({ id: 'foo', data: 'bar1' });
          server.close();
          client.send({ id: 'foo', data: 'bar2' });
        });
        client.on('error', err => err);
        client.on('close', () => console.log('close'));
        setTimeout(done, 5000);
      });
    });

    it('when receive error multiple times', done => {
      server.start(9090, () => {
        const client = new Client({
          host: '127.0.0.1',
          port: 9090,
        });

        client.once('connect', () => {
          server.close();
          const error = new Error('ECONNRESET');
          error.code = 'ECONNRESET';
          client._socket && client._socket.emit('error', error);
          client._socket && client._socket.emit('error', error);
        });
        client.on('error', err => err);
        client.on('close', () => console.log('close'));
        setTimeout(done, 5000);
      });
    });
  });
});


test/index.test.js
'use strict';

const mm = require('mm');
const os = require('os');
const path = require('path');
const TCPBase = require('../');
const assert = require('assert');
const pedding = require('pedding');
const { sleep } = require('mz-modules');
const server = require('./support/server');

describe('test/index.test.js', () => {
  class Client extends TCPBase {
    constructor(options) {
      Object.assign(options, {
        headerLength: 8,
        heartbeatInterval: 3000,
      });
      super(options);
    }

    getBodyLength(header) {
      return header.readInt32BE(0);
    }

    decode(buf, header) {
      let data;
      if (buf) {
        data = JSON.parse(buf);
      }
      return {
        id: header.readInt32BE(4),
        data,
      };
    }

    get heartBeatPacket() {
      return makeRequest(1).data;
    }
  }

  class Client2 extends Client {
    constructor(options) {
      Object.assign(options, {
        headerLength: 0,
        needHeartbeat: false,
      });

      super(options);
    }

    getHeader() {
      return this.read(8);
    }
  }

  class Client3 extends Client2 {
    decode() {
      throw new Error('mock error');
    }
  }

  function makeRequest(id, content, oneway) {
    const header = new Buffer(8);
    header.fill(0);
    const body = new Buffer(JSON.stringify(content || {
      id,
      message: 'hello',
    }));
    header.writeInt32BE(body.length, 0);
    header.writeInt32BE(id, 4);
    return {
      id,
      oneway,
      data: Buffer.concat([ header, body ]),
    };
  }

  let client;
  let client2;
  const sockPath = path.join(os.tmpdir(), 'tcp-base-test.sock');
  before(done => {
    server.start(12201, err => {
      if (err) {
        return done(err);
      }
      client = new Client({
        host: '127.0.0.1',
        port: 12201,
      });

      client.ready(() => {
        if (os.platform() === 'win32') {
          done();
        } else {
          server.start(sockPath, err => {
            if (err) {
              return done(err);
            }
            client2 = new Client({
              path: sockPath,
            });

            client2.ready(done);
          });
        }
      });
    });
  });

  afterEach(mm.restore);

  after(function* () {
    yield client.close();
    yield client.close();
    if (client2) {
      client2.close();
    }
    server.close();
  });

  it('client should be ok', () => {
    assert(client.isOK);
  });

  it('should get address ok', () => {
    assert(client.address === '127.0.0.1:12201');
  });

  it('should send request ok', done => {
    client.send(makeRequest(1), (err, res) => {
      assert.ifError(err);
      assert.deepEqual(res, {
        id: 1,
        message: 'hello',
      });
      done();
    });
  });

  if (client2) {
    it('client2 should be ok', () => {
      assert(client2.isOK);
    });
    it('client2 should send request ok', done => {
      client2.send(makeRequest(1001), (err, res) => {
        assert.ifError(err);
        assert.deepEqual(res, {
          id: 1001,
          message: 'hello',
        });
        done();
      });
    });
  }

  it('should send thunk ok', function* () {
    const res = yield client.sendThunk(makeRequest(2));
    assert.deepEqual(res, {
      id: 2,
      message: 'hello',
    });
  });

  it('should send oneway ok', function* () {
    yield client.sendThunk(makeRequest(3, {
      id: 3,
      noResponse: true,
    }, true));
    assert(client._invokes.size === 0);
  });

  it('should emit error if connect timeout', done => {
    const client = new Client({
      host: '127.0.0.1',
      port: 12000,
    });
    client.on('error', err => {
      assert(err);
      assert(err.message.includes('connect ECONNREFUSED 127.0.0.1:12000'));
      done();
    });
  });

  it('should emit close if socket is destroyed', done => {
    let client = new Client({
      host: '127.0.0.1',
      port: 12201,
    });

    client.on('close', () => {
      client = new Client({
        host: '127.0.0.1',
        port: 12201,
      });
    });

    client._socket.destroy();

    setTimeout(() => {
      client.send(makeRequest(1), (err, res) => {
        assert.ifError(err);
        assert.deepEqual(res, {
          id: 1,
          message: 'hello',
        });
        done();
      });
    }, 1000);
  });

  // it('should emit close in the same tick', function* () {
  //   let client = new Client({
  //     host: '127.0.0.1',
  //     port: 12201,
  //   });

  //   client.on('close', () => {
  //     client = new Client({
  //       host: '127.0.0.1',
  //       port: 12201,
  //     });
  //   });

  //   client.close();

  //   const res = yield client.sendThunk(makeRequest(2));
  //   assert.deepEqual(res, {
  //     id: 2,
  //     message: 'hello',
  //   });
  // });

  it('should emit error if socket has been closed', done => {
    const client = new Client3({
      host: '127.0.0.1',
      port: 12201,
    });

    client.close();

    client.send(makeRequest(1), err => {
      assert(err);
      assert(err.name === 'SocketCloseError');
      done();
    });
  });

  it('should emit error if parse header error', done => {
    done = pedding(2, done);

    const client = new Client3({
      host: '127.0.0.1',
      port: 12201,
    });

    client.on('error', err => {
      assert(err);
      assert(err.message.includes('mock error'));
      done();
    });

    client.send(makeRequest(1), err => {
      assert(err);
      assert(err.message.includes('mock error'));
      done();
    });
  });

  it('should emit error if parse body error', done => {
    done = pedding(2, done);

    const client = new Client3({
      host: '127.0.0.1',
      port: 12201,
    });

    client.on('error', err => {
      assert(err);
      assert(err.message.includes('mock error'));
      done();
    });

    client.send(makeRequest(1), err => {
      assert(err);
      assert(err.message.includes('mock error'));
      done();
    });
  });

  it('should process request timeout well', done => {
    client.send(makeRequest(4, {
      id: 4,
      noResponse: true,
    }, false), err => {
      assert(err);
      [ 'id', 'dataLength', 'bufferSize1', 'bufferSize2', 'startTime', 'endTime', 'writeSuccess' ]
        .forEach(p => {
          assert(err.socketMeta.hasOwnProperty(p));
        });
      assert(err.socketMeta.writeSuccess);
      assert(err.message === 'Server no response in 3000ms, address#127.0.0.1:12201');
      done();
    });
  });

  it('should wait for drain if buffer is full', function* () {
    const queue = [];
    const content = require('../package.json');
    for (let i = 5; i < 10000; i++) {
      const req = makeRequest(i, {
        id: i,
        content,
      });
      req.timeout = 10000;
      queue.push(client.sendThunk(req));
    }
    yield queue;
  });

  it('should clean all invoke if client is close', done => {
    const cli = new Client({
      host: '127.0.0.1',
      port: 12201,
    });
    cli.ready()
      .then(() => {
        cli.send(makeRequest(2, {
          id: 2,
          timeout: 2000,
        }), err => {
          assert(err);
          assert(err.name === 'SocketCloseError');
        });
        cli._queue.push([ makeRequest(1, {
          id: 1,
          timeout: 2000,
        }), err => {
          assert(err);
          assert(err.name === 'SocketCloseError');
        } ]);
        cli.close();
      })
      .catch(err => done(err));
    cli.on('close', () => {
      assert(cli._invokes.size === 0);
      assert(cli._queue.length === 0);
      done();
    });
  });

  it('should override getHeader ok', function* () {
    const client = new Client2({
      host: '127.0.0.1',
      port: 12201,
    });
    yield client.ready();
    const res = yield client.sendThunk(makeRequest(1));
    assert.deepEqual(res, {
      id: 1,
      message: 'hello',
    });
  });

  it('should send heartbeat request', function* () {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
    });
    yield client.ready();
    yield sleep(100);
    assert(client._heartbeatTimer);
    mm(client._socket, 'write', buf => {
      assert.deepEqual(buf, client.heartBeatPacket);
      client.emit('heartbeat');
    });
    yield client.await('heartbeat');
    yield client.close();
  });

  it('should override sendHeartBeat', function* () {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
    });

    mm(client, 'sendHeartBeat', () => {
      client.send(makeRequest(10), (err, res) => {
        client.emit('heartbeat', res);
      });
    });

    yield client.ready();
    yield sleep(100);

    assert(client._heartbeatTimer);
    const heartbeat = yield client.await('heartbeat');
    assert(heartbeat.id === 10);
    yield client.close();
  });

  it('should support concurrent', function* () {
    const concurrent = 3;

    class Client5 extends Client2 {
      constructor(options) {
        Object.assign(options, {
          concurrent,
        });
        super(options);
      }

      send(packet, callback) {
        assert(this._invokes.size <= concurrent);
        return super.send(packet, callback);
      }
    }

    const client = new Client5({
      host: '127.0.0.1',
      port: 12201,
    });
    const queue = [];
    for (let i = 0; i < 20000; i++) {
      const req = makeRequest(i);
      req.timeout = 10000;
      queue.push(client.sendThunk(req));
    }
    yield queue;
  });

  it('should send before ready', done => {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
    });
    client.send(makeRequest(1), (err, res) => {
      assert.ifError(err);
      assert.deepEqual(res, {
        id: 1,
        message: 'hello',
      });
      done();
    });
  });

  it('should process empty body ok', done => {
    const header = new Buffer(8);
    header.writeInt32BE(0, 0);
    header.writeInt32BE(1000, 4);
    const request = {
      id: 1000,
      oneway: false,
      data: header,
    };
    client.send(request, (err, res) => {
      assert.ifError(err);
      assert(!res);
      done();
    });
  });

  it('should not emit PacketParsedError  if message is handled with error in event listener', done => {
    const listeners = process.listeners('uncaughtException');
    process.removeAllListeners('uncaughtException');

    const handle = err => {
      assert(err.message === 'responseError');
      process.removeAllListeners('uncaughtException');
      listeners.forEach(listener => {
        process.on('uncaughtException', listener);
      });
      done();
    };
    process.on('uncaughtException', handle);

    let first = false;
    client.on('response', () => {
      if (!first) {
        first = true;
        throw new Error('responseError');
      }
    });

    client.send(makeRequest(1), err => {
      assert.ifError(err);
    });
  });

  it('send response packet', done => {
    mm(client.options, 'concurrent', 1);
    // 1. send 正常请求,
    // 2. 连续 send oneway 请求.
    // 3. 验证 queue 的数量和队列的内容是否符合预期.
    done = pedding(4, done);
    const c0 = {
      id: 10,
      message: 'hello',
      timeout: 100,
    };
    const c1 = {
      id: 11,
      message: 'hello',
      timeout: 500,
    };
    const c2 = {
      id: 12,
      message: 'hello',
      timeout: 1000,
    };
    client.send(makeRequest(10, c0), (err, res) => {
      assert.ifError(err);
      assert.deepEqual(res, c0);
      done();
    });
    client.send(makeRequest(11, c1), (err, res) => {
      assert.ifError(err);
      assert.deepEqual(res, c1);
      done();
    });
    setTimeout(() => {
      // c1 目前已经在 queue 里面, 然后连续添加2个 oneway.
      client.send(makeRequest(15, {
        id: 5,
        oneway: true,
      }, true));
      client.send(makeRequest(16, {
        id: 6,
        oneway: true,
      }, true));

      client.send(makeRequest(17), (err, res) => {
        assert.ifError(err);
        assert.deepEqual(res, { id: 17, message: 'hello' });
        done();
      });

      const queue = client._queue;
      assert(queue.length === 2);
      assert(queue[0][0].id === 12);
      assert(queue[1][0].id === 17);
    }, 600);

    client.send(makeRequest(12, c2), (err, res) => {
      assert.ifError(err);
      assert.deepEqual(res, c2);
      done();
    });
  });

  it('should not emit PacketParsedError if error occurred in callback function', done => {
    const listeners = process.listeners('uncaughtException');
    process.removeAllListeners('uncaughtException');

    const handle = err => {
      assert(err.message === 'callbackError');
      process.removeAllListeners('uncaughtException');
      listeners.forEach(listener => {
        process.on('uncaughtException', listener);
      });
      done();
    };
    process.on('uncaughtException', handle);

    client.on('error', () => {
      assert(false, 'should not run');
    });

    client.send(makeRequest(1), err => {
      assert.ifError(err);
      throw new Error('callbackError');
    });
  });

  it('should not emit Error if socket ECONNRESET', function* () {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
    });
    yield client.ready();
    const err = new Error('123');
    err.code = 'ECONNRESET';
    client._socket.on('error', () => {
      client._socket.destroy();
    });
    yield Promise.race([
      client.await('error'),
      client.await('close'),
      client._socket.emit('error', err),
    ]);
  });

  it('should handle connect timeout ok', function* () {
    const client = new Client({
      host: '1.1.1.1',
      port: 12200,
      connectTimeout: 300,
    });
    try {
      yield client.ready();
      assert(false, 'should not run here');
    } catch (err) {
      assert(err.name === 'TcpConnectionTimeoutError');
      assert(err.message.includes('[TCPBase] socket connect timeout (300ms)'));
    }
  });

  it('should get error if socket is closed', function* () {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
      connectTimeout: 300,
    });
    yield client.ready();
    yield client.close();
    try {
      yield client.sendThunk(makeRequest(1));
      assert(false, 'should not run here');
    } catch (err) {
      assert(err && err.message === '[TCPBase] The socket was closed. (address: 127.0.0.1:12201)');
    }
    try {
      yield Promise.race([
        client.await('error'),
        client.send(makeRequest(2, 'haha', true)),
      ]);
      assert(false, 'should not run here');
    } catch (err) {
      assert(err && err.message === '[TCPBase] The socket was closed. (address: 127.0.0.1:12201)');
      assert(err.oneway === true);
    }
  });
});


test/heartbeat.test.js
'use strict';

const net = require('net');
const TCPBase = require('../');
const awaitEvent = require('await-event');
const sleep = require('mz-modules/sleep');

describe('test/heartbeat.test.js', () => {
  let id = 0;

  function makeRequest(data) {
    id++;
    const header = new Buffer(8);
    header.fill(0);
    const body = new Buffer(data || '');
    header.writeInt32BE(body.length, 0);
    header.writeInt32BE(id, 4);
    return {
      id,
      data: Buffer.concat([ header, body ]),
    };
  }

  class Client extends TCPBase {
    constructor(options) {
      Object.assign(options, {
        headerLength: 8,
      });
      super(options);
    }

    getBodyLength(header) {
      return header.readInt32BE(0);
    }

    decode(buf, header) {
      let data;
      if (buf) {
        data = buf.toString();
      }
      return {
        id: header.readInt32BE(4),
        data,
      };
    }

    get heartBeatPacket() {
      return makeRequest('heartbeat').data;
    }
  }

  let server;
  before(function* () {
    server = net.createServer(socket => {
      socket.on('data', data => {
        const len = data.readInt32BE(0);
        // 正常请求 hold 住不返回，只返回 heartbeat
        if (len) {
          socket.write(data);
        }
      });
    });
    server.listen(12201);
    yield awaitEvent(server, 'listening');
  });

  after(function* () {
    server.close();
    yield awaitEvent(server, 'close');
  });

  it('should heartbeat ok', function* () {
    const client = new Client({
      host: '127.0.0.1',
      port: 12201,
      heartbeatInterval: 1000,
      maxIdleTime: 5000,
    });

    yield sleep(1200);

    const timer = setInterval(() => {
      client.send(makeRequest());
    }, 500);

    yield Promise.race([
      awaitEvent(client, 'error'),
      sleep(5000),
    ]);

    clearInterval(timer);
    client.close();
  });
});


example/client.js
'use strict';

const TCPBase = require('../lib/base');

/**
 * A Simple Protocol:
 *   (4B): request id
 *   (4B): body length
 *   ------------------------------
 *   body data
 */
class Client extends TCPBase {
  getHeader() {
    return this.read(8);
  }

  getBodyLength(header) {
    return header.readInt32BE(4);
  }

  decode(body, header) {
    return {
      id: header.readInt32BE(0),
      data: body,
    };
  }

  // heartbeat packet
  get heartBeatPacket() {
    return new Buffer([ 255, 255, 255, 255, 0, 0, 0, 0 ]);
  }
}

const client = new Client({
  host: '127.0.0.1',
  port: 8080,
});

const body = new Buffer('hello');
const data = new Buffer(8 + body.length);
data.writeInt32BE(1, 0);
data.writeInt32BE(body.length, 4);
body.copy(data, 8, 0);

client.send({
  id: 1,
  data,
  timeout: 5000,
}, (err, res) => {
  if (err) {
    console.error(err);
  }
  console.log(res.toString()); // should echo 'hello'
});


lib/base.js
'use strict';

const net = require('net');
const is = require('is-type-of');
const assert = require('assert');
const Base = require('sdk-base');

const addressKey = Symbol('address');
const defaultOptions = {
  noDelay: true,
  connectTimeout: 3000,
  responseTimeout: 3000,
  heartbeatInterval: 5000,
  needHeartbeat: true,
  concurrent: 0,
  logger: console,
};
const noop = () => {};
let seed = 0;

class TCPBase extends Base {
  /**
   * A base class for tcp client with basic functions
   *
   * @param {Object} options
   *   - {String} host - server host
   *   - {Number} port - server port
   *   - {Number} headerLength - length of the packet header, this field is optional,
   *                             but if you not provider, you must override getHeader method
   *   - {Boolean} [noDelay] - whether use the Nagle algorithm or not，defaults to true
   *   - {Number} [concurrent] - the number of concurrent packet, defaults to zero, means no limit
   *   - {Number} [responseTimeout] - limit the maximum time for waiting a response
   *   - {Logger} [logger] - the logger client
   * @class
   */
  constructor(options) {
    super();

    this.options = Object.assign({}, defaultOptions, options);
    if (!this.options.path) {
      assert(this.options.host, 'options.host is required');
      assert(this.options.port, 'options.port is required');
    }

    if (this.options.needHeartbeat) {
      assert(this.heartBeatPacket, 'heartBeatPacket getter must be implemented if needHeartbeat');
    }

    this.clientId = ++seed;
    this._heartbeatTimer = null;
    this._socket = null;
    this._header = null;
    this._bodyLength = null;
    this._lastError = null;
    this._queue = [];
    this._invokes = new Map();
    this[addressKey] = this.options.host + ':' + this.options.port;
    this._lastHeartbeatTime = 0;
    this._lastReceiveDataTime = 0;

    this._connect();
    this.ready(err => {
      if (!err && this.options.needHeartbeat) {
        this._startHeartbeat();
      }
    });
  }

  /**
   * get packet header
   *
   * @return {Buffer} header
   */
  getHeader() {
    return this.read(this.options.headerLength);
  }

  /* eslint-disable valid-jsdoc, no-unused-vars */

  /**
   * get body length from header
   *
   * @param {Buffer} header - header data
   * @return {Number} bodyLength
   */
  getBodyLength(header) {
    throw new Error('not implement');
  }

  /**
   * return a heartbeat packet
   *
   * @property {Buffer} TCPBase#heartBeatPacket
   */
  get heartBeatPacket() {
    throw new Error('not implement');
  }

  /**
   * send heartbeat packet
   *
   * @return {void}
   */
  sendHeartBeat() {
    this._socket.write(this.heartBeatPacket);
  }

  /**
   * deserialze method, leave it to implement by subclass
   *
   * @param {Buffer} buf - binary data
   * @return {Object} packet object
   */
  decode(buf) {
    throw new Error('not implement');
  }

  /* eslint-enable valid-jsdoc, no-unused-vars */

  /**
   * if the connection is writable, also including flow control logic
   *
   * @property {Boolean} TCPBase#_writable
   */
  get _writable() {
    if (this.options.concurrent && this._invokes.size >= this.options.concurrent) {
      return false;
    }

    return this.isOK;
  }

  /**
   * if the connection is healthy or not
   *
   * @property {Boolean} TCPBase#isOK
   */
  get isOK() {
    return this._socket && this._socket.writable;
  }

  /**
   * remote address
   *
   * @property {String} TCPBase#address
   */
  get address() {
    return this[addressKey];
  }

  /**
   * logger
   *
   * @property {Logger} TCPBase#logger
   */
  get logger() {
    return this.options.logger;
  }

  /**
   * Pulls some data out of the socket buffer and returns it.
   * If no data available to be read, null is returned
   *
   * @param {Number} n - to specify how much data to read
   * @return {Buffer} - data
   */
  read(n) {
    return this._socket.read(n);
  }

  /**
   * send packet to server
   *
   * @param {Object} packet
   *   - {Number} id - packet id
   *   - {Buffer} data - binary data
   *   - {Boolean} [oneway] - oneway or not
   *   - {Number} [timeout] - the maximum time for waiting a response
   * @param {Function} [callback] - Call this function，when processing is complete, optional.
   * @return {void}
   */
  send(packet, callback = noop) {
    if (!this._socket) {
      const err = new Error(`[TCPBase] The socket was closed. (address: ${this[addressKey]})`);
      err.id = packet.id;
      err.data = packet.data.toString('base64');
      if (packet.oneway) {
        err.oneway = true;
        callback();
        this.emit('error', err);
      } else {
        callback(err);
      }
      return;
    }
    if (packet.oneway) {
      this._socket.write(packet.data);
      callback();
      return;
    }
    if (!this._writable) {
      this._queue.push([ packet, callback ]);
      return;
    }
    const meta = {
      id: packet.id,
      dataLength: packet.data.length,
      bufferSize1: this._socket.bufferSize,
      bufferSize2: -1,
      startTime: Date.now(),
      endTime: -1,
    };
    let endTime;
    meta.writeSuccess = this._socket.write(packet.data, () => {
      endTime = Date.now();
    });
    const timeout = packet.timeout || this.options.responseTimeout;
    this._invokes.set(packet.id, {
      meta,
      packet,
      timer: setTimeout(() => {
        if (!this._socket) {
          return;
        }

        meta.bufferSize2 = this._socket.bufferSize;
        meta.endTime = endTime;
        this._finishInvoke(packet.id);
        const err = new Error(`Server no response in ${timeout}ms, address#${this[addressKey]}`);
        err.socketMeta = meta;
        err.name = 'ResponseTimeoutError';
        callback(err);
      }, timeout),
      callback,
    });
  }

  /**
   * thunk style api of send(packet, callback)
   *
   * @param {Object} packet
   *   - {Number} id - packet id
   *   - {Buffer} data - binary data
   *   - {Boolean} [oneway] - oneway or not
   *   - {Number} [timeout] - the maximum time for waiting a response
   * @return {Function} thunk function
   */
  sendThunk(packet) {
    return callback => this.send(packet, callback);
  }

  _finishInvoke(id) {
    this._invokes.delete(id);
    if (this._writable) {
      this._resume();
    }
  }

  _errorCallback(callback, err) {
    if (!err) {
      err = new Error(`The socket was closed. (address: ${this[addressKey]})`);
      err.name = 'SocketCloseError';
    }
    callback && callback(err);
  }

  // mark all invokes timeout
  _cleanInvokes(err) {
    for (const id of this._invokes.keys()) {
      const req = this._invokes.get(id);
      clearTimeout(req.timer);
      this._errorCallback(req.callback, err);
    }
    this._invokes.clear();
  }

  // clean up the queue
  _cleanQueue(err) {
    let args = this._queue.pop();
    while (args) {
      // args[0] 是packet， args[1]是callback
      this._errorCallback(args[1], err);
      args = this._queue.pop();
    }
  }

  _resume() {
    const args = this._queue.shift();
    if (args) {
      this.send(args[0], args[1]);
    }
  }

  // read data from socket，and decode it to packet object
  _readPacket() {
    if (is.nullOrUndefined(this._bodyLength)) {
      this._header = this.getHeader();
      if (!this._header) {
        return false;
      }
      this._bodyLength = this.getBodyLength(this._header);
    }

    let body;
    if (this._bodyLength > 0) {
      body = this.read(this._bodyLength);
      if (!body) {
        return false;
      }
    }
    this._bodyLength = null;
    const entity = this.decode(body, this._header);
    // the schema of entity
    // {
    //   id: 'request id',
    //   isResponse: true,
    //   data: {}  // deserialized object
    // }
    let type = 'request';
    if (!entity.hasOwnProperty('isResponse')) {
      entity.isResponse = this._invokes.has(entity.id);
    }
    if (entity.isResponse) {
      type = 'response';
      const invoke = this._invokes.get(entity.id);
      if (invoke) {
        this._finishInvoke(entity.id);
        clearTimeout(invoke.timer);
        process.nextTick(() => {
          invoke.callback(entity.error, entity.data);
        });
      }
    }
    if (entity.data) {
      process.nextTick(() => {
        this.emit(type, entity, this[addressKey]);
      });
    }
    return true;
  }

  /**
   * close the socket
   *
   * @param {Error} err - the error which makes socket closed
   * @return {void}
   */
  close(err) {
    if (!this._socket) {
      return Promise.resolve();
    }
    this._socket.destroy(err);
    return this.await('close');
  }

  _handleClose() {
    if (!this._socket) {
      return;
    }
    this._socket.removeAllListeners();
    this._socket = null;

    this._cleanInvokes(this._lastError);
    // clean timer
    if (this._heartbeatTimer) {
      clearInterval(this._heartbeatTimer);
      this._heartbeatTimer = null;
    }
    this._cleanQueue(this._lastError);
    this.emit('close');
  }

  _handleReadable() {
    this._lastReceiveDataTime = Date.now();
    try {
      let remaining = false;
      do {
        remaining = this._readPacket();
      } while (remaining);
    } catch (err) {
      this.close(err);
    }
  }

  _connect(done) {
    if (!done) {
      done = err => {
        this.ready(err ? err : true);
      };
    }
    const { port, host, localAddress, localPort, family, hints, lookup, path } = this.options;
    const socket = this._socket = net.connect({
      port, host, localAddress, localPort, family, hints, lookup, path,
    });
    socket.setNoDelay(this.options.noDelay);
    socket.on('readable', () => { this._handleReadable(); });
    socket.once('close', () => { this._handleClose(); });
    socket.on('error', err => {
      err.message += ' (address: ' + this[addressKey] + ')';
      this._lastError = err;
      if (err.code === 'ECONNRESET') {
        this.logger.warn('[TCPBase] socket is closed by other side while there were still unhandled data in the socket buffer');
      } else {
        this.emit('error', err);
      }
    });
    socket.setTimeout(this.options.connectTimeout, () => {
      const err = new Error(`[TCPBase] socket connect timeout (${this.options.connectTimeout}ms)`);
      err.name = 'TcpConnectionTimeoutError';
      err.host = this.options.host;
      err.port = this.options.port;
      this.close(err);
    });

    socket.once('connect', () => {
      // set timeout back to zero after connected
      socket.setTimeout(0);
      this.emit('connect');
    });

    Promise.race([
      this.await('connect'),
      this.await('error'),
    ]).then(done, done);
  }

  _startHeartbeat() {
    this._heartbeatTimer = setInterval(() => {
      const duration = this._lastHeartbeatTime - this._lastReceiveDataTime;
      if (this._lastReceiveDataTime && duration > this.options.heartbeatInterval) {
        const err = new Error(`server ${this[addressKey]} no response in ${duration}ms, maybe the socket is end on the other side.`);
        err.name = 'ServerNoResponseError';
        this.close(err);
        return;
      }
      // flow control
      if (this._invokes.size > 0 || !this.isOK) {
        return;
      }
      this._lastHeartbeatTime = Date.now();
      this.sendHeartBeat();
    }, this.options.heartbeatInterval);
  }
}

module.exports = TCPBase;
